contracts/marketpools.js

```diff
@@ -0,0 +1,348 @@
+  uPool.basePrice = api.BigNumber(uPool.quoteQuantity).dividedBy(uPool.baseQuantity).toFixed(pool.precision);
+  uPool.quotePrice = api.BigNumber(uPool.baseQuantity).dividedBy(uPool.quoteQuantity).toFixed(pool.precision);
+  if (swap) {
+    uPool.baseVolume = api.BigNumber(uPool.baseVolume).plus(Math.abs(baseAdjusted)).toFixed(pool.precision);
+    uPool.quoteVolume = api.BigNumber(uPool.quoteVolume).plus(Math.abs(quoteAdjusted)).toFixed(pool.precision);
+  }
+  await api.db.update('pools', uPool);
+}
+
+actions.create = async (payload) => {
+  const {
+    tokenPair, isSignedWithActiveKey,
+  } = payload;
+
+  // get contract params
+  const params = await api.db.findOne('params', {});
+  const { poolCreationFee } = params;
+
+  // eslint-disable-next-line no-template-curly-in-string
+  const utilityTokenBalance = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'" });
+
+  const authorizedCreation = api.BigNumber(poolCreationFee).lte(0) || api.sender === api.owner
+    ? true
+    : utilityTokenBalance && api.BigNumber(utilityTokenBalance.balance).gte(poolCreationFee);
+
+  if (api.assert(authorizedCreation, 'you must have enough tokens to cover the creation fee')
+    && await validateTokenPair(tokenPair)
+    && api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')) {
+    const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+    const baseToken = await api.db.findOneInTable('tokens', 'tokens', { symbol: baseSymbol });
+    const quoteToken = await api.db.findOneInTable('tokens', 'tokens', { symbol: quoteSymbol });
+    const newPool = {
+      tokenPair,
+      baseQuantity: 0,
+      baseVolume: 0,
+      basePrice: 0, // Base per Quote (usual way of reading a pair)
+      quoteQuantity: 0,
+      quoteVolume: 0,
+      quotePrice: 0, // Quote per Base (reverse price)
+      precision: Math.min(baseToken.precision, quoteToken.precision),
+      active: true,
+      creator: api.sender,
+    };
+    await api.db.insert('pools', newPool);
+
+    // burn the token creation fees
+    if (api.sender !== api.owner && api.BigNumber(poolCreationFee).gt(0)) {
+      await api.executeSmartContract('tokens', 'transfer', {
+        // eslint-disable-next-line no-template-curly-in-string
+        to: 'null', symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'", quantity: poolCreationFee, isSignedWithActiveKey,
+      });
+    }
+    api.emit('createPool', { tokenPair });
+  }
+};
+
+actions.addLiquidity = async (payload) => {
+  const {
+    tokenPair,
+    baseQuantity,
+    quoteQuantity,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    || !api.assert(baseQuantity && api.BigNumber(baseQuantity).gt(0), 'invalid baseQuantity')
+    || !api.assert(quoteQuantity && api.BigNumber(quoteQuantity).gt(0), 'invalid quoteQuantity')
```

Comment:
You should validate that baseQuantity and quoteQuantity are within the precision limits for each token type.

Reply:
Fixed